File: src/FastSlowPointersTraversal/141_LinkedListCycle/LinkedListCycle.cs
// Hash table 184ms

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public bool HasCycle(ListNode head) {
        HashSet<ListNode> visited = new HashSet<ListNode>();
        while (head != null)
        {
            visited.Add(head);
            if (visited.Contains(head.next))
            {                    
                return true;
            }
            head = head.next;
        }
        return false;
    }
}

/*
Time complexity: O(n).
The algorithm makes one traversal of the list head, there are n operations and time complexity is O(n).

Space complexity: O(n)
Dictionary will store each elemnt of head.
*/


// slow fast pointers 96ms

public class Solution {
    public bool HasCycle(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;
        while(slow != null && fast != null && fast.next != null)
        {
            fast = fast.next.next;
            slow = slow.next;
            
            if(slow == fast) return true;
        }        
        return false;
    }
}

/*
Time complexity: O(n).
The algorithm makes one traversal of the list head, there are n operations and time complexity is O(n).

Space complexity: O(1)
extra instant space.

*/

File: src/FastSlowPointersTraversal/141_LinkedListCycle/LinkedListCycle.py
# slow fast pointers

# Runtime: 56 ms, faster than 63.56% of Python3 online submissions for Linked List Cycle.
# Memory Usage: 17.7 MB, less than 56.41% of Python3 online submissions for Linked List Cycle.

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def hasCycle(self, head) -> bool:
        fast = head
        slow = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            if fast == slow:
                return True
        return False

    if __name__ == '__main__':
        print('hello')
        

# Time complexity: O(n).v
# The algorithm makes one traversal of the list head, there are n operations and time complexity is O(n).

# Space complexity: O(1)
# extra instant space.
